File: 3. Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.go
// 3. Longest Substring Without Repeating Characters (5/27/2021)
// Runtime: 8 ms (69.28%) Memory: 2.82 MB (66.36%) 

func lengthOfLongestSubstring(s string) int {
	solution := 0

	dict := make(map[byte]int)

	front, back := 0, 0
	for front = 0; front < len(s); front++ {
		if pos, exists := dict[s[front]]; exists {
			for back <= pos {
				delete(dict, s[back])
				back++
			}
		} else if front-back+1 > solution {
			solution = front - back + 1
		}
		dict[s[front]] = front
	}

	return solution
}

File: 3. Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.js
// 3. Longest Substring Without Repeating Characters (2/16/53726)
// Runtime: 88 ms (86.92%) Memory: 43.28 MB (94.98%) 

/**
 * @param {string} s
 * @return {number}
 */
function lengthOfLongestSubstring(s) {
    let start = 0,
        end = 0,
        max = 0;
    const set = new Set();
    while (end < s.length) {
        if (!set.has(s[end])) {
            set.add(s[end++]);
            max = Math.max(max, set.size);
        } else set.delete(s[start++]);
    }
    return max;
}